src/main/java/org/spongepowered/api/command/Command.java

```diff
@@ -0,0 +1,535 @@
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.api.command;
+
+import com.google.common.collect.Lists;
+import org.spongepowered.api.Sponge;
+import org.spongepowered.api.command.exception.CommandException;
+import org.spongepowered.api.command.managed.ChildExceptionBehavior;
+import org.spongepowered.api.command.managed.ChildExceptionBehaviors;
+import org.spongepowered.api.command.managed.CommandExecutor;
+import org.spongepowered.api.command.manager.CommandManager;
+import org.spongepowered.api.command.manager.CommandMapping;
+import org.spongepowered.api.command.parameter.Parameter;
+import org.spongepowered.api.command.parameter.flag.Flags;
+import org.spongepowered.api.command.parameter.token.InputTokenizer;
+import org.spongepowered.api.event.cause.Cause;
+import org.spongepowered.api.plugin.PluginContainer;
+import org.spongepowered.api.text.Text;
+import org.spongepowered.api.util.ResettableBuilder;
+import org.spongepowered.api.world.Location;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.function.BiFunction;
+import java.util.function.BiPredicate;
+
+import javax.annotation.Nullable;
+
+/**
+ * The Command interface is the low-level interface that all commands in the
+ * Sponge ecosystem inherit.
+ *
+ * <p><strong>Most</strong> plugins are highly recommended (but not obligated)
+ * to use {@link Command#builder()} to create commands. The
+ * {@link Command.Builder} allows plugins to take advantage of a higher level
+ * of abstraction, such as argument parsers and simple child command handling,
+ * removing the need for boilerplate code.</p>
+ *
+ * <p>Plugins are free to implement this interface should they prefer to do so.
+ * Custom implementations of this class are not required to implement a sane
+ * {@link Object#equals(Object)}, but are highly encouraged to do so.</p>
+ *
+ * <p>Commands in Sponge are provided with a {@link Cause}, as well as an
+ * associated {@link CommandSource}. Note that the {@link CommandSource} is
+ * not necessarily the object that actually invoked the command, however, it
+ * is the source that is <strong>responsible</strong> for the command. Therefore,
+ * in general, command eligibility checks should be performed against the
+ * provided {@link CommandSource}. However, users and implementors may act upon
+ * the {@link Cause} alone as they see fit, should it be preferred.</p>
+ */
+public interface Command {
```

Comment:
I disagree with this name, as I said before - this name should be reserved for annotation-based commands, when they come.

Reply:
**Absolutely not.** I don't see a compelling reason as to why this should be renamed. Why should this, which is a command and nothing more, be displaced by an annotation which builds on top of this? It's completely backwards.

A `@Command` annotation can still exist in a different package, done right, this interface won't need to be implemented in the same class as the annotation being used.

Reply:
Sure, but how about cases where you'd have both types in a single class? Fully-qualified imports suck.

Feel free to suggest an alternative name for the annotation.

Reply:
Sure, they do. I'll cross that bridge when we come to it.

Reply:
A good name for the annotation might be `@CommandDefinition`

Reply:
the term I use as well as Mojang is RootCommand for the /root portion, then subcommands for nodes underneath that.

In ACF I named it @CommandAlias but @RootCommand would of been nice, but I already had that used as an interface.